### ISSUE TITLE
Permit return of @MonotonicNonNull field

### ISSUE THREAD

**Issue by mernst** [typetools/checker-framework#2216]
An assignment of `@MonotonicNonNull` to `@MonotonicNonNull` is forbidden, because the two might be in different states.
However, it is acceptable when a method has a single `return` statement.  Consider the following code, which currently does not type-check but should:

```java
import org.checkerframework.checker.nullness.qual.MonotonicNonNull;

public class MonotonicNonNullReturn {

  @MonotonicNonNull Object field;

  @MonotonicNonNull Object getter() {
    return field;
  }

}
```

A less restrictive rule would be that if the lhs has only one assignment (statically), then the assignment is permitted (even though lhs and rhs are both `@MonotonicNonNull`).  When a routine has only one `return` statement, this is satisfied.

**Comment by the1derer** [comment]
@mernst Is anyone working on this?
If not I would like to solve this issue as this will greatly help in solving classes with no Constructor and only getter and setter method without having to use ``@RequiresNonNull`` which can cause contracts.precondition errors.
Can you give me any additional information which would help in solving this.

**Comment by wmdietl** [comment]
What is the motivation for allowing `@MonotonicNonNull` on a return type?
Isn't it always equivalent to making the return type `@Nullable`?
It would help to describe the updates to the documentation of the qualifier to understand what the proposed semantics are. At the moment it focuses only on fields (and sometimes variables) but doesn't mention return types.
Instead of `@MonotonicNonNull` for return types one should use `@Nullable` and for parameter types one should use `@NonNull`.

**Comment by mernst** [comment]
> What is the motivation for allowing `@MonotonicNonNull` on a return type?
> Isn't it always equivalent to making the return type `@Nullable`?

Not for `@Pure` methods.  Yes, for non-pure methods.

**Comment by mernst** [comment]
> Is anyone working on this?

If no one is assigned to it, then no one is working on it.

> Can you give me any additional information which would help in solving this.

I can't answer this question without knowing what you have tried so far or what your specific questions are.

**Comment by the1derer** [comment]
> I can't answer this question without knowing what you have tried so far or what your specific questions are.

Take the following example:
```java
import org.checkerframework.checker.nullness.qual.MonotonicNonNull;
import org.checkerframework.checker.nullness.qual.RequiresNonNull;
import org.checkerframework.checker.nullness.qual.EnsuresNonNull;

interface A {
  
  public String getName();
}

class MonotonicNonNullReturn implements A {

  @MonotonicNonNull String name;

  @EnsuresNonNull("this.name")
  public void setName(String name) {
    this.name=name;
  }

  @RequiresNonNull("name")
  @Override
  public String getName() {
    return name;
  }

}
```
Command
``javacheck -processor nullness FILENAME.java``
Here I am receiving following error
```log
Test.java:21: error: [contracts.precondition.override.invalid] Precondition must be no stronger than in superclass.
  public String getName() {
                ^
  Method
    @Initialized @NonNull String getName(@Initialized @NonNull MonotonicNonNullReturn this) in MonotonicNonNullReturn
    cannot override
    @Initialized @NonNull String getName(@Initialized @NonNull A this) in A
  Precondition is too strong: '@NonNull' for 'this.name'
1 error
```
I can't annotate getName() of A as there can be no field ``name``.So, In this case having return type ``@MonotonicNonNull`` could be helpful.
